applications/StructuralMechanicsApplication/custom_response_functions/adjoint_structural_solver.py

```diff
@@ -0,0 +1,156 @@
+                node.AddDof(StructuralMechanicsApplication.ADJOINT_ROTATION_X)
+                node.AddDof(StructuralMechanicsApplication.ADJOINT_ROTATION_Y)
+                node.AddDof(StructuralMechanicsApplication.ADJOINT_ROTATION_Z)
+        self.print_on_rank_zero("::[AdjointMechanicalSolver]:: ", "DOF's ADDED.")
+
+    def Initialize(self):
+        """Perform initialization after adding nodal variables and dofs to the main model part. """
+
+        domain_size = self.main_model_part.ProcessInfo[KratosMultiphysics.DOMAIN_SIZE]
+        if self.response_function_settings["response_type"].GetString() == "adjoint_local_stress":
+            if (domain_size == 2):
+                raise Exception("Currently only availible for 3D. Your choice is 2D")
+            elif (domain_size == 3):
+                self.response_function = StructuralMechanicsApplication.AdjointLocalStressResponseFunction(self.main_model_part, self.response_function_settings)
+            else:
+                raise Exception("Invalid DOMAIN_SIZE: " + str(domain_size))
+        elif self.response_function_settings["response_type"].GetString() == "adjoint_nodal_displacement":
+            if (domain_size == 2):
+                raise Exception("Currently only availible for 3D. Your choice is 2D")
+            elif (domain_size == 3):
+                self.response_function = StructuralMechanicsApplication.AdjointNodalDisplacementResponseFunction(self.main_model_part, self.response_function_settings)
+            else:
+                raise Exception("Invalid DOMAIN_SIZE: " + str(domain_size))
+        elif self.response_function_settings["response_type"].GetString() == "adjoint_strain_energy":
+            if (domain_size == 2):
+                raise Exception("Currently only availible for 3D. Your choice is 2D")
+            elif (domain_size == 3):
+                self.response_function = StructuralMechanicsApplication.AdjointStrainEnergyResponseFunction(self.main_model_part, self.response_function_settings)
+            else:
+                raise Exception("Invalid DOMAIN_SIZE: " + str(domain_size))
+        else:
+            raise Exception("invalid response_type: " + self.response_function_settings["response_type"].GetString())
+
+        super(AdjointStructuralSolver, self).Initialize()
+
+        self.print_on_rank_zero("::[AdjointMechanicalSolver]:: ", "Finished initialization.")
+
+    def Solve(self):
+        if self.response_function_settings["response_type"].GetString() == "adjoint_strain_energy":
+            self._SolveSpecialStrainEnergy()
+        else:
+            if self.settings["clear_storage"].GetBool():
+                self.Clear()
+            mechanical_solution_strategy = self.get_mechanical_solution_strategy()
+            mechanical_solution_strategy.Solve()
+
+    def _SolveSpecialStrainEnergy(self):
+
+        self.response_function.Initialize()
+
+        for node in self.main_model_part.Nodes:
```

Comment:
I would have done this in vector form (all the DISPLACEMENTs at once.) your call though